File: src-tauri/src/task.rs
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Local};
use rand::Rng;

#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum State {
    Pending,
    Doing,
    Paused,
    Canceld,
    Done,
    Failed,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct TaskConfig {
    pub title: String,
    pub state: State,
    pub parent: u64,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct Task {
    pub id: u64,
    pub title: String,
    pub create_time: Option<DateTime<Local>>,
    pub last_update_time: Option<DateTime<Local>>,
    pub parent: u64,
    pub childs: Vec<u64>,
    pub state: State,
}

impl Task {
    pub fn from_config(cfg: TaskConfig) -> Self {
        return Task {
            id: alloc_id(),
            title: cfg.title,
            create_time: Some(Local::now()),
            last_update_time: Some(Local::now()),
            parent: cfg.parent,
            childs: Vec::new(),
            state: cfg.state
        }
    }

}


fn alloc_id() -> u64 {
    return rand::thread_rng().gen_range(1..10001);
}

File: src/task/task.ts
import { invoke } from '@tauri-apps/api'

export interface Task {
    id: number,
    title: string,
    create_time?: Date | null,
    last_update_time?: Date | null,
    parent: number,
    childs: [number] | [],
    state: "Pending" | "Doing" | "Paused" | "Canceld" | "Done" | "Failed"
  }
  
  export interface TaskConfig {
    title: string,
    state: "Pending" | "Doing" | "Paused" | "Canceld" | "Done" | "Failed"
    parent: number
  }
  
  export async function create_task_with_title(title: string) :Promise<number> {
      console.log("title is ", title)
      return await invoke("create_task", {
        t: {
          title: title,
        } as TaskConfig,
      })
  }
  export async function create_task(task_cfg: TaskConfig) :Promise<number> {
      console.log("create task with cfg:", task_cfg)
      return await invoke("create_task", {t: task_cfg})
  }
  
  export async function get_task(id: number) :Promise<Task|null> {
      return await invoke("get_task", {
        id: id,
      })
  }
  
  export async function list_tasks() :Promise<[Task]> {
      return await invoke("list_task", {})
  }